core/trino-main/src/main/java/io/trino/operator/project/PageFieldsToInputParametersRewriter.java

```diff
@@ -42,80 +45,117 @@ private PageFieldsToInputParametersRewriter() {}
+                            .map(expression -> expression.accept(this, unconditionallyEvaluated))
                             .collect(toImmutableList()));
         }
 
         @Override
-        public RowExpression visitSpecialForm(SpecialForm specialForm, Void context)
+        public RowExpression visitSpecialForm(SpecialForm specialForm, Boolean unconditionallyEvaluated)
         {
+            if (unconditionallyEvaluated) {
+                switch (specialForm.getForm()) {
+                    case IF:
+                    case SWITCH:
+                    case BETWEEN:
+                    case AND:
+                    case OR:
+                    case COALESCE:
+                        return specialFormWithConditionallyEvaluatedArguments(specialForm);
+                    default:
+                        break;
+                }
+            }
             return new SpecialForm(
                     specialForm.getForm(),
                     specialForm.getType(),
                     specialForm.getArguments().stream()
-                            .map(expression -> expression.accept(this, context))
+                            .map(expression -> expression.accept(this, unconditionallyEvaluated))
                             .collect(toImmutableList()),
                     specialForm.getFunctionDependencies());
         }
 
         @Override
-        public RowExpression visitConstant(ConstantExpression literal, Void context)
+        public RowExpression visitConstant(ConstantExpression literal, Boolean unconditionallyEvaluated)
         {
             return literal;
         }
 
         @Override
-        public RowExpression visitLambda(LambdaDefinitionExpression lambda, Void context)
+        public RowExpression visitLambda(LambdaDefinitionExpression lambda, Boolean unconditionallyEvaluated)
         {
             return new LambdaDefinitionExpression(
                     lambda.getArgumentTypes(),
                     lambda.getArguments(),
-                    lambda.getBody().accept(this, context));
+                    lambda.getBody().accept(this, unconditionallyEvaluated));
         }
 
         @Override
-        public RowExpression visitVariableReference(VariableReferenceExpression reference, Void context)
+        public RowExpression visitVariableReference(VariableReferenceExpression reference, Boolean unconditionallyEvaluated)
         {
             return reference;
         }
+
+        private SpecialForm specialFormWithConditionallyEvaluatedArguments(SpecialForm specialForm)
+        {
+            List<RowExpression> arguments = specialForm.getArguments();
+            return new SpecialForm(
+                    specialForm.getForm(),
+                    specialForm.getType(),
+                    IntStream.range(0, arguments.size()).boxed()
+                            // All the arguments after the first one are assumed to be conditionally evaluated
```

Comment:
This is very brittle. We should treat each function separately and encode their assumptions explicitly

Reply:
I've explicitly enumerated all the cases in visitSpecialForm now